rfd/0076-Track-Monthly-Active-Users.md

```diff
@@ -0,0 +1,221 @@
+	var user string
+
+	if g, ok := event.(UserMetadataGetter); ok {
+		user = g.GetUser()
+	}
+
+	return user
+}
+```
+
+With this, we can modify the `EmitAuditEvent` in Teleport Enterprise so that it extracts the user and records it as active (the `ActiveUsers` struct below is detailed in the [next subsection](#track-active-users-in-the-backend)):
+```go
+// EmitAuditEvent emits the specified event.
+func (l *AuditLog) EmitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
+	l.setUserAsActive(ctx, event)
+	return trace.Wrap(l.Inner.EmitAuditEvent(ctx, event))
+}
+
+// setUserAsActive tries to extract a user from the event and sets it as active in case it's found.
+func (l *AuditLog) setUserAsActive(ctx context.Context, event apievents.AuditEvent) {
+	user := events.GetUser(event)
+	if user != "" {
+		err := l.ActiveUsers.SetAsActive(ctx, user)
+		l.Log.WithError(err).Error("Failed to set user as active.")
+	}
+}
+```
+
+As we detail in the [next subsection](#track-active-users-in-the-backend), `ActiveUsers.SetAsActive` makes a call to the backend, slowing down this codepath responsible for emitting audit events.
+If performance is a concern here, we can call `ActiveUsers.SetAsActive` asynchronously.
+
+### Track active users in the backend
+
+In order to track and count the number of active users, we provide an `ActiveUsers` struct with two methods: `SetAsActive` and `Count`.
+
+The `SetAsActive` method receives the `user` and upserts the key `metrics/active_users/USER` into the backend using [`Backend.Put`](https://github.com/gravitational/teleport/blob/8a27614b83590056e0d43394b926cf6db29b190b/lib/backend/backend.go#L46-L48) with a 30-day TTL.
+Independently of whether the user has been active in the last 30 days or not, this ensures that the user is considered as active during the next 30 days.
+
+```go
+// SetAsActive registers a user as active
+func (a *ActiveUsers) SetAsActive(ctx context.Context, user string) error {
+	_, err := a.Backend.Put(ctx, a.backendItem(user))
+	return trace.Wrap(err)
+}
+
+func (a *ActiveUsers) backendItem(user string) backend.Item {
+	// item expires in 30 days
+	ttl := 30 * 24 * time.Hour
+	now := a.Clock.Now().UTC()
+	return backend.Item{
+		Key:     activeUsersKey(user),
+		Value:   []byte(now.Format(time.RFC3339)),
+		Expires: now.Add(ttl),
+	}
+}
+
+func activeUsersKey(user string) []byte {
+	return backend.Key("metrics", "active_users", user)
+}
+```
+
+The `ActiveUsers` struct also provides a `Count` method that computes the number of active users during the last 30 days.
+This method leverages the [`Backend.GetRange`](https://github.com/gravitational/teleport/blob/8a27614b83590056e0d43394b926cf6db29b190b/lib/backend/backend.go#L60-L61) method to count the number of keys with the `metrics/active_users` prefix.
+
+```go
+// Count counts the number of active users during the last 30 days
+func (a *ActiveUsers) Count(ctx context.Context) (int, error) {
```

Comment:
I think this we should move this logic and calculation out of the core backend and into the Postgres. This will split the load of running analytical and range queries from the core DB.